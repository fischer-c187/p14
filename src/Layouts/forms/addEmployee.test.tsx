import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { beforeEach, describe, expect, it, vi } from "vitest";
import AddEmployee from "./addEmployee";

const mocks = vi.hoisted(() => ({
  usePostEmployee: vi.fn(() => ({
    mutate: vi.fn(),
    isPending: false,
  })),
  useFormValidation: vi.fn(() => ({
    errors: {},
    register: vi.fn(() => {}),
    validateForm: vi.fn(() => true),
  })),
}));

vi.mock("@hooks/usePostEmployee", () => ({
  default: mocks.usePostEmployee,
}));

vi.mock("@hooks/useFormValidation", () => ({
  default: mocks.useFormValidation,
}));

describe("AddEmployee", () => {
  beforeEach(() => {
    vi.resetModules();
  });
  it("should match the snapshot", () => {
    const queryClient = new QueryClient();
    const handleSuccessSubmission = vi.fn();
    const handleFailedSubmission = vi.fn();
    const { container } = render(
      <QueryClientProvider client={queryClient}>
        <AddEmployee
          onSuccessfulSubmission={handleSuccessSubmission}
          onFailedSubmission={handleFailedSubmission}
        />
      </QueryClientProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should display the loader when the form is submitting", () => {
    vi.mocked(mocks.usePostEmployee).mockReturnValue({
      mutate: vi.fn(),
      isPending: true,
    });

    const queryClient = new QueryClient();
    const handleSuccessSubmission = vi.fn();
    const handleFailedSubmission = vi.fn();
    render(
      <QueryClientProvider client={queryClient}>
        <AddEmployee
          onSuccessfulSubmission={handleSuccessSubmission}
          onFailedSubmission={handleFailedSubmission}
        />
      </QueryClientProvider>
    );
    expect(screen.getByTestId("loader")).toBeInTheDocument();
  });

  it("calls onSuccessfulSubmission after successful mutation", async () => {
    vi.mocked(mocks.usePostEmployee).mockReturnValue({
      mutate: vi.fn((_, { onSuccess, onError }) =>
        Promise.resolve()
          .then(() => (onSuccess ? onSuccess("data de réponse simulée") : null))
          .catch((error) => (onError ? onError(error) : null))
      ),
      isPending: false,
    });

    vi.mock("@hooks/useFormValidation", () => ({
      default: vi.fn(() => ({
        errors: {},
        register: vi.fn(() => {}),
        validateForm: vi.fn(() => true),
      })),
    }));

    const queryClient = new QueryClient();
    const handleSuccessSubmission = vi.fn();
    const handleFailedSubmission = vi.fn();

    const user = userEvent.setup();

    render(
      <QueryClientProvider client={queryClient}>
        <AddEmployee
          onSuccessfulSubmission={handleSuccessSubmission}
          onFailedSubmission={handleFailedSubmission}
        />
      </QueryClientProvider>
    );
    await user.type(screen.getByLabelText(/First Name/i), "John");
    await user.type(screen.getByLabelText(/Last Name/i), "Doe");
    await user.type(screen.getByLabelText(/Birth of Date/i), "01/01/1990");
    await user.type(screen.getByLabelText(/Start Date/i), "01/01/2022");
    await user.type(screen.getByLabelText(/Street/i), "123 Main St");
    await user.type(screen.getByLabelText(/City/i), "Anytown");
    // Pour les dropdowns, vous pouvez simuler un choix comme ceci :
    await user.type(screen.getByLabelText(/Zip Code/i), "12345");

    await user.click(screen.getByRole("button", { name: /save/i }));
    await waitFor(() => {
      expect(handleSuccessSubmission).toHaveBeenCalledTimes(1);
    });
  });

  it("calls onFailedSubmission after failed mutation", async () => {
    vi.mocked(mocks.usePostEmployee).mockReturnValue({
      mutate: vi.fn((_, { onSuccess, onError }) =>
        Promise.reject(new Error("Erreur simulée"))
          .then(() => (onSuccess ? onSuccess("data de réponse simulée") : null))
          .catch((error) => (onError ? onError(error) : null))
      ),
      isPending: false,
    });

    vi.mock("@hooks/useFormValidation", () => ({
      default: vi.fn(() => ({
        errors: {},
        register: vi.fn(() => {}),
        validateForm: vi.fn(() => true),
      })),
    }));

    const queryClient = new QueryClient();
    const handleSuccessSubmission = vi.fn();
    const handleFailedSubmission = vi.fn();

    const user = userEvent.setup();

    render(
      <QueryClientProvider client={queryClient}>
        <AddEmployee
          onSuccessfulSubmission={handleSuccessSubmission}
          onFailedSubmission={handleFailedSubmission}
        />
      </QueryClientProvider>
    );
    await user.type(screen.getByLabelText(/First Name/i), "John");
    await user.type(screen.getByLabelText(/Last Name/i), "Doe");
    await user.type(screen.getByLabelText(/Birth of Date/i), "01/01/1990");
    await user.type(screen.getByLabelText(/Start Date/i), "01/01/2022");
    await user.type(screen.getByLabelText(/Street/i), "123 Main St");
    await user.type(screen.getByLabelText(/City/i), "Anytown");
    // Pour les dropdowns, vous pouvez simuler un choix comme ceci :
    await user.type(screen.getByLabelText(/Zip Code/i), "12345");

    await user.click(screen.getByRole("button", { name: /save/i }));
    await waitFor(() => {
      expect(handleFailedSubmission).toHaveBeenCalledTimes(1);
    });
  });
});
