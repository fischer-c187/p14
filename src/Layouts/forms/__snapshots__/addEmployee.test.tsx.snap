// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddEmployee > should match the snapshot 1`] = `
<div>
  <form
    class="space-y-8 m-auto lg:w-2/3  min-w-80  w-5/6"
  >
    <div
      class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
      data-testid="form-raw"
    >
      <div
        class="flex flex-col gap-2 w-full"
        data-testid="formFieldWithLabel"
      >
        <label
          class="text-gray-700 font-medium text-sm w-fit"
          for="firstName"
        >
          First Name
        </label>
        <input
          class="border py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-citron-600 focus:border-transparent"
          id="firstName"
          name="firstName"
          placeholder="John"
          required=""
          type="text"
        />
      </div>
      <div
        class="flex flex-col gap-2 w-full"
        data-testid="formFieldWithLabel"
      >
        <label
          class="text-gray-700 font-medium text-sm w-fit"
          for="lastName"
        >
          Last Name
        </label>
        <input
          class="border py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-citron-600 focus:border-transparent"
          id="lastName"
          name="lastName"
          placeholder="Doe"
          required=""
          type="text"
        />
      </div>
    </div>
    <div
      class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
      data-testid="form-raw"
    >
      <div
        class="flex flex-col gap-2 w-full"
        data-testid="formFieldWithLabel"
      >
        <label
          class="text-gray-700 font-medium text-sm w-fit"
          for="birthDate"
        >
          Birth of Date
        </label>
        <div
          class="_inputWrapper_rkpbb_1"
          data-testid="datepicker"
        >
          <input
            class="border w-full py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-hrnet-green focus:border-transparent"
            data-testid="dateInput"
            id="birthDate"
            name="birthDate"
            placeholder="MM/dd/YYYY"
            required=""
            tabindex="0"
            title="Enter the date in this format: MM/DD/YYYY or MM-DD-YYYY"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="flex flex-col gap-2 w-full"
        data-testid="formFieldWithLabel"
      >
        <label
          class="text-gray-700 font-medium text-sm w-fit"
          for="startDate"
        >
          Start Date
        </label>
        <div
          class="_inputWrapper_rkpbb_1"
          data-testid="datepicker"
        >
          <input
            class="border w-full py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-hrnet-green focus:border-transparent"
            data-testid="dateInput"
            id="startDate"
            name="startDate"
            placeholder="MM/dd/YYYY"
            required=""
            tabindex="0"
            title="Enter the date in this format: MM/DD/YYYY or MM-DD-YYYY"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <fieldset
      class="pt-0 pb-8 space-y-8 bg-gray-200 border-y-4 border-gray-300"
    >
      <legend
        class="text-lg font-medium ml-4 bg-hrnet-green text-white px-2"
      >
        Address
      </legend>
      <div
        class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
        data-testid="form-raw"
      >
        <div
          class="flex flex-col gap-2 w-full"
          data-testid="formFieldWithLabel"
        >
          <label
            class="text-gray-700 font-medium text-sm w-fit"
            for="street"
          >
            Street
          </label>
          <input
            class="border py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-citron-600 focus:border-transparent"
            id="street"
            name="street"
            placeholder="123 Main St."
            type="text"
          />
        </div>
        <div
          class="flex flex-col gap-2 w-full"
          data-testid="formFieldWithLabel"
        >
          <label
            class="text-gray-700 font-medium text-sm w-fit"
            for="city"
          >
            City
          </label>
          <input
            class="border py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-citron-600 focus:border-transparent"
            id="city"
            name="city"
            placeholder="New York"
            type="text"
          />
        </div>
      </div>
      <div
        class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
        data-testid="form-raw"
      >
        <div
          class="flex flex-col gap-2 w-full"
          data-testid="formFieldWithLabel"
        >
          <label
            class="text-gray-700 font-medium text-sm w-fit"
            for="state"
          >
            State
          </label>
          <div
            class="relative"
            data-testid="dropdownWrapper"
            role="button"
            tabindex="-1"
          >
            <div
              class="flex items-center justify-between rounded-md border px-5 py-2 shadow-sm border-gray-300 bg-white focus:border-transparent focus:outline-none focus:ring-2 focus:ring-citron-600 group-focus:focus:ring-citron-600 text-gray-500"
              data-testid="dropdownBtn"
              role="button"
              tabindex="0"
            >
              <span
                class="mr-4"
              >
                Alabama
              </span>
              <img
                alt="icon"
                class="fill-gray-500 transition-transform"
                src="/src/assets/angle-up-solid.svg"
              />
            </div>
            <input
              class="sr-only"
              id="state"
              name="state"
              type="text"
              value="alabama"
            />
          </div>
        </div>
        <div
          class="flex flex-col gap-2 w-full"
          data-testid="formFieldWithLabel"
        >
          <label
            class="text-gray-700 font-medium text-sm w-fit"
            for="zipCode"
          >
            Zip Code
          </label>
          <input
            class="border py-2 px-3 rounded-md  border-gray-300 focus:outline-none focus:ring-2 focus:ring-citron-600 focus:border-transparent"
            id="zipCode"
            name="zipCode"
            pattern="^\\d{5}(-\\d{4})?$"
            placeholder="New York"
            type="text"
          />
        </div>
      </div>
    </fieldset>
    <div
      class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
      data-testid="form-raw"
    >
      <div
        class="flex flex-col gap-2 w-full row-span-2"
        data-testid="formFieldWithLabel"
      >
        <label
          class="text-gray-700 font-medium text-sm w-fit"
          for="department"
        >
          Department
        </label>
        <div
          class="relative"
          data-testid="dropdownWrapper"
          role="button"
          tabindex="-1"
        >
          <div
            class="flex items-center justify-between rounded-md border px-5 py-2 shadow-sm border-gray-300 bg-white focus:border-transparent focus:outline-none focus:ring-2 focus:ring-citron-600 group-focus:focus:ring-citron-600 text-gray-500"
            data-testid="dropdownBtn"
            role="button"
            tabindex="0"
          >
            <span
              class="mr-4"
            >
              Sales
            </span>
            <img
              alt="icon"
              class="fill-gray-500 transition-transform"
              src="/src/assets/angle-up-solid.svg"
            />
          </div>
          <input
            class="sr-only"
            id="department"
            name="department"
            type="text"
            value="sales"
          />
        </div>
      </div>
    </div>
    <div
      class="grid px-4 lg:px-8 justify-start gap-8 grid-rows-2 grid-cols-1 md:grid-rows-1 md:grid-cols-2"
      data-testid="form-raw"
    >
      <button
        class="items-center bg-hrnet-green col-span-2 row-span-2 justify-self-center mt-4 text-white py-2 px-4 rounded-md hover:bg-hrnet-green-600"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
